File: programovanie/builder/07grafika/opengl/priklady/bmp/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
#include <gl\gl.h>// Hlavièkový soubor pro OpenGL32 knihovnu
#include <gl\glu.h>// Hlavièkový soubor pro Glu32 knihovnu
#include <gl\glaux.h>// Hlavièkový soubor pro Glaux knihovnu
#include <soil\soil.h>

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
char bmp[256][256][3];
float s=0,t=0;
int uholx,uholy=20;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{    ogl_inicialized=0;
     ogl_init();
}
void TForm1::ogl_draw()
{ glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
//  glRotatef(20.0,1.0f,0.0f,0.0f);
//  glRotatef(uholy,0.0f,1.0f,0.0f);

//tu zadávame príkazy na kreslenie
  glColor4f(1.0f, 1.0f, 1.0f,1.0f);
    glBegin(GL_QUADS);
         glTexCoord2f(0.0, 0.0); glVertex3f(-6, +4, 0);
        glTexCoord2f(2.0, 0.0); glVertex3f(+6, +4, 0);
        glTexCoord2f(2.0*10/12, 1.0); glVertex3f(+4, -4, 0);
        glTexCoord2f(2.0*2/12, 1.0); glVertex3f(-4, -4, 0);
glEnd();

glFlush();
SwapBuffers(hdc);

}

//---------------------------------------------------------------------------

void __fastcall TForm1::destroy(TObject *Sender)
{
  ogl_exit();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::resize(TObject *Sender)
{
ogl_resize();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
for (int y=0;y<256;y++)             // pro vsechny radky pixmapy
        for (int x=0;x<256;x++)        // pro vsechny pixely na radku
        if ((y/32+x/32)%2)
            bmp[y][x][0]= bmp[y][x][1]= bmp[y][x][2]= 0;// cierne policko
        else
		{ //  bmp[y][x][0]=256-abs(x-128)-abs(y-128); bmp[y][x][1]=(256-abs(x-128)-abs(y-128))/2; bmp[y][x][2]= abs(x-128)+abs(y-128);
                    bmp[y][x][0]=x;  bmp[y][x][1]=y; bmp[y][x][2]=0;}
    glEnable(GL_TEXTURE_2D);       // povolime texturu
 //   glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);      // v smere osi x povolime opakovanie textury
 //   glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);      // v smere osi y povolime opakovanie textury
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);  // spôsob výpoètu ïalších bodov (textura vacsia)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexImage2D(GL_TEXTURE_2D, 0, 3, 256, 256, 0, GL_RGB, GL_UNSIGNED_BYTE, &bmp);

}
//---------------------------------------------------------------------------
void __fastcall TForm1::Timer1Timer(TObject *Sender)
{  uholy=(uholy+1)%360;//uholx=(uholx+1)%360;
   ogl_draw();
}
//---------------------------------------------------------------------------

File: programovanie/builder/07grafika/opengl/ukazky/textura2/unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs,opengl,glaux, StdCtrls, ExtCtrls;

type
  TForm1 = class(TForm)
    Button1: TButton;
    Timer1: TTimer;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure Timer1Timer(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;
  RC: HGLRC=0;
  DC: HDC=0;
  tex : gluInt;
  uholy:integer=0;

PROCEDURE glGenTextures(n: GLsizei; VAR textures: GLuint); STDCALL; EXTERNAL opengl32;
PROCEDURE glBindTexture(target: GLenum; texture: GLuint); STDCALL; EXTERNAL  opengl32;
implementation

{$R *.dfm}

procedure TForm1.FormCreate(Sender: TObject);
var
  PFD: TPixelFormatDescriptor;
  PF: Integer;
  textur: PTAUX_RGBImageRec;
begin
  // Nastavenie Device Contextu:
  DC:=GetDC(Handle);
  FillChar(PFD, SizeOf(TPixelFormatDescriptor), 0);
  PFD.nSize:=SizeOf(TPixelFormatDescriptor);
  PFD.nVersion:=1;
  PFD.dwFlags:=PFD_DRAW_TO_WINDOW or PFD_SUPPORT_OPENGL or PFD_DOUBLEBUFFER;
  PFD.iPixelType:=PFD_TYPE_RGBA;
  PFD.cColorBits:=24;
  PFD.cDepthBits:=32;
  PFD.iLayerType:=PFD_MAIN_PLANE;
  PF:=ChoosePixelFormat(DC, @PFD);
  if PF=0 then Exit;
  SetPixelFormat(DC, PF, @PFD);
  // Vytvorenie Rendering Contextu:
  RC:=wglCreateContext(DC);
  wglMakeCurrent(DC, RC);
  glClearColor(0, 0, 0, 0);
  glClearDepth(1);
  glEnable(GL_DEPTH_TEST);
  //zapneme priesvitnosù
  glEnable(GL_BLEND);
  glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

  glViewport(0, 0, ClientWidth, ClientHeight);
  glMatrixMode(GL_PROJECTION);
  glLoadIdentity;
  glOrtho(-8,+8,-8,+8,-8,+8);

 glenable(gl_texture_2d); //Texturen aktivieren
  textur := auxDIBImageLoadA('t1-128x128.bmp');
if not Assigned( textur ) then begin showmessage('Fehler beim Laden der Textur'); close;End;
glGenTextures(1, tex);
glBindTexture(GL_TEXTURE_2D, tex);
glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MAG_FILTER,GL_linear);
glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MIN_FILTER,GL_linear);
glTexImage2D(GL_TEXTURE_2D, 0, 3, textur^.sizeX, textur^.sizeY, 0, GL_RGB, GL_UNSIGNED_BYTE, textur^.data);
//glLoadIdentity;
end;


procedure TForm1.FormDestroy(Sender: TObject);
begin
  wglMakeCurrent(0, 0);
  wglDeleteContext(RC);
  RC:=0;
  ReleaseDC(Handle, DC);
  DC:=0;

end;

procedure TForm1.Button1Click(Sender: TObject);
begin
glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT);
glMatrixMode( GL_TEXTURE  );
glLoadIdentity;
glBindTexture(GL_TEXTURE_2D, tex); //Hiermit w‰hlen wir die Textur an
glMatrixMode(GL_MODELVIEW);
glLoadIdentity;

glbegin(gl_quads);
//glColor3f(1.0,1.0,0.2);
gltexcoord2f(0,0); //UV ñ Zuweisung f¸r den ersten Punkt
glvertex3f(-5,-5,-3);
gltexcoord2f(0,1); //UV ñ Zuweisung f¸r den zweiten Punkt
glvertex3f(-5, 5,-3);
gltexcoord2f(1,1); //UV ñ Zuweisung f¸r den dritten Punkt
glvertex3f( 5, 5,-3);
gltexcoord2f(1,0); //UV ñ Zuweisung f¸r den vierten Punkt
glvertex3f( 5,-5,-3);
glend;
SwapBuffers(DC);
end;

procedure TForm1.Timer1Timer(Sender: TObject);
begin
uholy:=(uholy+2) mod 360;
glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT);
glMatrixMode(GL_MODELVIEW);
glLoadIdentity;
glRotate( uholy,0,1,0);
glbegin(gl_quads);
//glColor3f(1.0,1.0,0.2);
gltexcoord2f(0,0); //UV ñ Zuweisung f¸r den ersten Punkt
glvertex3f(-5,-5,-3);
gltexcoord2f(0,1); //UV ñ Zuweisung f¸r den zweiten Punkt
glvertex3f(-5, 5,-3);
gltexcoord2f(1,1); //UV ñ Zuweisung f¸r den dritten Punkt
glvertex3f( 5, 5,-3);
gltexcoord2f(1,0); //UV ñ Zuweisung f¸r den vierten Punkt
glvertex3f( 5,-5,-3);
glend;
SwapBuffers(DC);

end;

end.
 